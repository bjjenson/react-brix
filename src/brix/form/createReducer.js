import { fromJS, Map } from 'immutable'
import { useReducer, useMemo } from 'react'
import { fieldReducer, initState, actions } from './fieldReducer'
import { defaultTextValue } from './fields/useTextField'
import { defaultSelectValue } from './fields/useSelectField'
import { defaultNumberValue } from './fields/useNumberField'
import { defaultBooleanValue } from './fields/useBooleanField'

export const createReducer = ({ fields, initialValues = Map() }) => {
  const derivedState = getInitialState(fields, initialValues)
  const [state, dispatch] = useReducer(fieldReducer, derivedState, initState)

  useMemo(() => {
    dispatch(actions.reset(derivedState))
    return initialValues
  }, [initialValues])

  return [state, dispatch]
}

/**
 * @param fields {Array<import("../form").IFormFieldArgs>}
 */
const getInitialState = (fields, initialValues) => {
  const fieldMap = fields.reduce((acc, field) => {
    return acc.set(field.name, generateDefaultFieldState(field, initialValues))
  }, Map())

  return fieldMap
}

export const generateDefaultFieldState = (field, initialValues) => {
  if (!field.name) throw new Error('name is required on field')

  const value = getInitialValue(field.value, initialValues.get(field.name), field.type)


  return fromJS({
    initial: {
      type: field.type || 'text',
      value,
      optional: field.optional || false,
      label: field.label,
    },
    current: {
      helperText: field.helperText || '',
      error: false,
      pristine: true,
      touched: false,
      value,
    },
  }).setIn(['initial', 'field'], field)
}

const defaultValues = {
  text: defaultTextValue,
  select: defaultSelectValue,
  number: defaultNumberValue,
  boolean: defaultBooleanValue,
}

const getInitialValue = (fieldValue, initialValue, type = 'text') => {
  if (initialValue) {
    return initialValue
  }
  if (fieldValue) {
    return fieldValue
  }
  return defaultValues[type]
}
