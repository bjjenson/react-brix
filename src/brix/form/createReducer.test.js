import { fromJS } from 'immutable'
import { useReducer, useMemo } from 'react'
import { createReducer } from './createReducer'

jest.mock('react')

let state, fields, initialValues, dispatch
beforeEach(() => {
  dispatch = 'dispatch'
  state = fromJS({
    first: {
      initial: {
        label: 'First',
      },
    },
  })
  fields = [
    { name: 'first', label: 'First' },
    { name: 'second', label: 'Second', type: 'boolean' },
  ]
  initialValues = fromJS({
    first: 'firstValue',
    second: 'secondValue',
  })

  useReducer.mockReturnValue([state, dispatch])
})

test('prepares initialState by field', () => {
  expect(createReducer({ fields, initialValues })).toMatchSnapshot()
  expect(useReducer.mock.calls[0]).toMatchSnapshot()
})

test('throws error if name not provided', () => {
  fields.push({ label: 'Third' })
  expect(() => createReducer({ fields, initialValues })).toThrowErrorMatchingSnapshot()
})

test('adds helperText from field config', () => {
  fields[0].helperText = 'i am helper'
  createReducer({ fields, initialValues })
  expect(useReducer.mock.calls[0]).toMatchSnapshot()
})

test('sets optional on field', () => {
  fields[0].optional = true
  createReducer({ fields, initialValues })
  expect(useReducer.mock.calls[0]).toMatchSnapshot()
})

describe('initialValues', () => {
  test('field.value should used if set', () => {
    fields = [
      { name: 'first', type: 'text', value: 'i am field value' },
    ]
    createReducer({ fields })
    expect(useReducer.mock.calls[0]).toMatchSnapshot()
  })

  test('text', () => {
    fields = [
      { name: 'first', type: 'text' },
    ]
    createReducer({ fields })
    expect(useReducer.mock.calls[0]).toMatchSnapshot()
  })

  test('select', () => {
    fields = [
      { name: 'first', type: 'select' },
    ]
    createReducer({ fields })
    expect(useReducer.mock.calls[0]).toMatchSnapshot()
  })

  test('number', () => {
    fields = [
      { name: 'first', type: 'number' },
    ]
    createReducer({ fields })
    expect(useReducer.mock.calls[0]).toMatchSnapshot()
  })

  test('boolean', () => {
    fields = [
      { name: 'first', type: 'boolean' },
    ]
    createReducer({ fields })
    expect(useReducer.mock.calls[0]).toMatchSnapshot()
  })

})

test('uses memo to check for initialValues change', () => {
  createReducer({ fields, initialValues })
  expect(useMemo.mock.calls[0]).toMatchSnapshot()
})

test('useMemo update dispatches reset', () => {
  let memoized
  dispatch = jest.fn()
  useMemo.mockImplementation((f) => {
    memoized = f
  })
  useReducer.mockReturnValue([state, dispatch])

  createReducer({ fields, initialValues })
  memoized()

  expect(dispatch.mock.calls[0]).toMatchSnapshot()
})
