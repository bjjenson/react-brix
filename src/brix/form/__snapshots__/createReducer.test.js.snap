// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds helperText from field config 1`] = `
Array [
  [Function],
  Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "value": "firstValue",
        "optional": false,
        "label": "First",
        "field": Object {
          "helperText": "i am helper",
          "label": "First",
          "name": "first",
        },
      },
      "current": Immutable.Map {
        "helperText": "i am helper",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "firstValue",
      },
    },
    "second": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "value": "secondValue",
        "optional": false,
        "label": "Second",
        "field": Object {
          "label": "Second",
          "name": "second",
          "type": "boolean",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "secondValue",
      },
    },
  },
  [Function],
]
`;

exports[`initialValues boolean 1`] = `
Array [
  [Function],
  Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "value": false,
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "boolean",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": false,
      },
    },
  },
  [Function],
]
`;

exports[`initialValues field.value should used if set 1`] = `
Array [
  [Function],
  Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "value": "i am field value",
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "text",
          "value": "i am field value",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "i am field value",
      },
    },
  },
  [Function],
]
`;

exports[`initialValues number 1`] = `
Array [
  [Function],
  Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "number",
        "value": 0,
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "number",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": 0,
      },
    },
  },
  [Function],
]
`;

exports[`initialValues select 1`] = `
Array [
  [Function],
  Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "select",
        "value": "",
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "select",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "",
      },
    },
  },
  [Function],
]
`;

exports[`initialValues text 1`] = `
Array [
  [Function],
  Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "value": "",
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "text",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "",
      },
    },
  },
  [Function],
]
`;

exports[`prepares initialState by field 1`] = `
Array [
  Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "label": "First",
      },
    },
  },
  "dispatch",
]
`;

exports[`prepares initialState by field 2`] = `
Array [
  [Function],
  Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "value": "firstValue",
        "optional": false,
        "label": "First",
        "field": Object {
          "label": "First",
          "name": "first",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "firstValue",
      },
    },
    "second": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "value": "secondValue",
        "optional": false,
        "label": "Second",
        "field": Object {
          "label": "Second",
          "name": "second",
          "type": "boolean",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "secondValue",
      },
    },
  },
  [Function],
]
`;

exports[`sets optional on field 1`] = `
Array [
  [Function],
  Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "value": "firstValue",
        "optional": true,
        "label": "First",
        "field": Object {
          "label": "First",
          "name": "first",
          "optional": true,
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "firstValue",
      },
    },
    "second": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "value": "secondValue",
        "optional": false,
        "label": "Second",
        "field": Object {
          "label": "Second",
          "name": "second",
          "type": "boolean",
        },
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "secondValue",
      },
    },
  },
  [Function],
]
`;

exports[`throws error if name not provided 1`] = `"name is required on field"`;

exports[`useMemo update dispatches reset 1`] = `
Array [
  Object {
    "payload": Immutable.Map {
      "first": Immutable.Map {
        "initial": Immutable.Map {
          "type": "text",
          "value": "firstValue",
          "optional": false,
          "label": "First",
          "field": Object {
            "label": "First",
            "name": "first",
          },
        },
        "current": Immutable.Map {
          "helperText": "",
          "error": false,
          "pristine": true,
          "touched": false,
          "value": "firstValue",
        },
      },
      "second": Immutable.Map {
        "initial": Immutable.Map {
          "type": "boolean",
          "value": "secondValue",
          "optional": false,
          "label": "Second",
          "field": Object {
            "label": "Second",
            "name": "second",
            "type": "boolean",
          },
        },
        "current": Immutable.Map {
          "helperText": "",
          "error": false,
          "pristine": true,
          "touched": false,
          "value": "secondValue",
        },
      },
    },
    "type": "reset",
  },
]
`;

exports[`uses memo to check for initialValues change 1`] = `
Array [
  [Function],
  Array [
    Immutable.Map {
      "first": "firstValue",
      "second": "secondValue",
    },
  ],
]
`;
